### PR TITLE
Versioning directives policy

### PR DESCRIPTION
**Pull request by story645**
Added policy to coding guidelines and PR guidelines based on discussion in #23506, samples of the directives are at  #24160

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -8,11 +8,14 @@
 - [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
 
 **Documentation**
-- [ ] New features are documented, with examples if plot related.
-- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
-- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
+- [ ] New plotting related features are documented with examples.
```
Comment (timhoffm) [inline .github/PULL_REQUEST_TEMPLATE.md:11]: This misses now a general statement "New features are documented".

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -8,11 +8,14 @@
...
 - [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
 
+**Release Notes**
+- [ ] New features noted with `versionadded` directive in docstring and documented in `doc/users/next_whats_new/`
```
Comment (timhoffm) [inline .github/PULL_REQUEST_TEMPLATE.md:15]:
```suggestion
- [ ] New features are marked with `.. versionadded::` directive in docstring and documented in `doc/users/next_whats_new/`
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -8,11 +8,14 @@
...
+- [ ] API changes noted with `versionchanged` directive in docstring and documented in `doc/api/next_api_changes/`
```
Comment (timhoffm) [inline .github/PULL_REQUEST_TEMPLATE.md:16]:
```suggestion
- [ ] API changes are marked with `.. versionchanged::` directive in docstring and documented in `doc/api/next_api_changes/`
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -22,7 +25,7 @@ consider the following points:
 - Help with git and github is available at
   https://matplotlib.org/devel/gitwash/development_workflow.html.
 
-- Do not create the PR out of main, but out of a separate branch.
+- Create the PR on a new branch, please do not make changes on `main`.
```
Comment (timhoffm) [inline .github/PULL_REQUEST_TEMPLATE.md:28]:
```suggestion
- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.
```

File: doc/devel/coding_guide.rst [hunk 5]
```diff
@@ -108,13 +108,60 @@ Documentation
...
+New features and API changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When adding a major new feature or changing the API in a backward incompatible
+way, please document it by including a versioning directive in the docstring
+and adding an entry to the folder for either the what's new or API change notes.
+
++-------------------+-----------------------------+----------------------------------+
+| for this addition | include this directive      | create entry in this folder      |
++===================+=============================+==================================+
+| new feature       | ``.. versionadded:: v3.N``  | :file:`doc/users/whats_new.rst`  |
```
Comment (timhoffm) [inline doc/devel/coding_guide.rst:122]:
```suggestion
| new feature       | ``.. versionadded:: v3.N``  | :file:`doc/users/next_whats_new/` |
```

**Comment by tacaswell** [comment]
Responding to a question from @jklymak in #24160, I think we should be on the liberal side of adding these notes.  It seems like we get ~ a bug report per month of "Why doesn't this documented feature work?!" to which the answer is "it was added in a newer version than you have installed.  If the docs clearly say what version the feature was added in hopefully we can reduce that user frustration.

That said, having the version switcher on the docs does help make it easier to see the docs for the version you have installed.

If we get to a point where we have so many of these where it is distracting, my knee-jerk thought is to start removing 3-4 version of markers to keep the total number down...

**Comment by jklymak** [comment]
That's fine with me. Otoh we need to be explicit about what is new if we do this for api changes.

**Comment by oscargus** [comment]
Should we also have some removal policy? I guess the documents may be a bit unreadable if version info for every change since 2004 would be in there...

**Comment by timhoffm** [comment]
I'd go with Tom's suggestion. We should start thinking of removal, when it get's too distracting. But that can be decided when we're there. Guessing up-front is not necessary.

**Comment by oscargus** [comment]
Agreed. Missed that comment though...

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Since this is not only documentation but sets a policy, I'd like to have a second review.

**Comment by timhoffm** [comment]
@story645 I've replied here: https://github.com/matplotlib/matplotlib/pull/24160#issuecomment-1286481707

File: doc/devel/coding_guide.rst [hunk 6]
```diff
@@ -108,13 +108,60 @@ Documentation
...
+New features and API changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When adding a major new feature or changing the API in a backward incompatible
+way, please document it by including a versioning directive in the docstring
+and adding an entry to the folder for either the what's new or API change notes.
+
++-------------------+-----------------------------+----------------------------------+
+| for this addition | include this directive      | create entry in this folder      |
++===================+=============================+==================================+
+| new feature       | ``.. versionadded:: v3.N``  | :file:`doc/users/next_whats_new/`|
```
Comment (QuLogic) [inline doc/devel/coding_guide.rst:122]: I don't think the `v` is necessary according to the Sphinx docs:
```suggestion
| new feature       | ``.. versionadded:: 3.N``   | :file:`doc/users/next_whats_new/`|
```

File: doc/devel/coding_guide.rst [hunk 7]
```diff
@@ -108,13 +108,60 @@ Documentation
...
++-------------------+-----------------------------+----------------------------------+
+| API change        | ``.. versionchanged:: v3.N``| :file:`doc/api/next_api_changes/`|
```
Comment (QuLogic) [inline doc/devel/coding_guide.rst:124]:
```suggestion
| API change        | ``.. versionchanged:: 3.N`` | :file:`doc/api/next_api_changes/`|
```

**Comment by timhoffm** [comment]
> Um should there be a placeholder letter instead of 3 too or not worth it for now?

This is clear enough. Let's not overthink this. (IMHO people would understand all of `X.Y`, `3.N`, `3.6`)

**Comment by timhoffm** [comment]
I'm not in favor of `major.minor.patch` because that is quite verbose and it's less clear that these are variables.

If anything I have a slight preference for the concrete example `.. versionadded:: 3.6`. - People are able to generalize from that and won't copy literally. Generalizing seems a bit simpler that guessing what a variable should be. But as said

> Let's not overthink this. (IMHO people would understand all of `X.Y`, `3.N`, `3.6`)

**Review by QuLogic — APPROVED** [review]

**Pull request merged**